src/main/scala/in/ashwanthkumar/suuchi/router/HandleOrForwardRouter.scala

```diff
@@ -17,26 +17,30 @@ class HandleOrForwardRouter(routingStrategy: RoutingStrategy, self: MemberAddres
   private val log = LoggerFactory.getLogger(getClass)
 
   override def interceptCall[ReqT, RespT](serverCall: ServerCall[ReqT, RespT], headers: Metadata, next: ServerCallHandler[ReqT, RespT]): Listener[ReqT] = {
-    log.trace("Intercepting " + serverCall.getMethodDescriptor.getFullMethodName + " method in " + self)
+    log.trace("Intercepting " + serverCall.getMethodDescriptor.getFullMethodName + " method in " + self + ", headers= " + headers.toString)
     new Listener[ReqT] {
       val delegate = next.startCall(serverCall, headers)
       var forwarded = false
 
       override def onReady(): Unit = delegate.onReady()
       override def onMessage(incomingRequest: ReqT): Unit = {
-        // TODO - Handle forwarding loop here
-        if(routingStrategy.route.isDefinedAt(incomingRequest)) {
-          routingStrategy route incomingRequest match {
-            case items if items.nonEmpty && !items.exists(_.equals(self)) =>
-              val node = items.head
-              log.trace(s"Forwarding request to ${node}")
-              val clientResponse: RespT = forward(serverCall, incomingRequest, node)
+        if (routingStrategy.route.isDefinedAt(incomingRequest)) {
+          val eligibleNodes = routingStrategy route incomingRequest
+          // Always set ELIGIBLE_NODES header to the list of nodes eligible in the current
+          // operation - as defined by the RoutingStrategy
+          headers.put(Metadata.Key.of(Headers.ELIGIBLE_NODES, MemberAddressMarshaller), eligibleNodes)
+
+          eligibleNodes match {
+            case nodes if nodes.nonEmpty && !nodes.exists(_.equals(self)) =>
+              val destination = nodes.head
+              log.trace(s"Forwarding request to $destination")
+              val clientResponse: RespT = forward(serverCall.getMethodDescriptor, headers, incomingRequest, destination)
```

Comment:
Today yeah - NettyTransport la. I think it would go through the full network stack.